server/app/views/FileUploadViewStrategy.java

```diff
@@ -44,18 +52,89 @@ public abstract class FileUploadViewStrategy extends ApplicationBaseView {
+                        BaseStyles.FORM_ERROR_TEXT_BASE,
+                        Styles.HIDDEN));
+    // TODO(#2589): Consider removing isSample entirely since it's only used for this
+    // codepath. The only change would be that extraFileUploadFields would need to
+    // allow an empty signedFileUploadRequest. Presumably this wouldn't be a large issue
+    // since uploads wouldn't work at runtime without the parameters.
+    if (!params.isSample()) {
+      result.with(extraFileUploadFields(params.signedFileUploadRequest().get()));
+    }
+    return result;
+  }
+
+  protected abstract ImmutableList<Tag> extraFileUploadFields(StorageUploadRequest request);
 
   /**
-   * Method to render the submit form for uploading a file.
+   * Method to render the UI for uploading a file.
    *
    * @param params the information needed to render a file upload view
    * @param applicantQuestionRendererFactory a class for rendering applicant questions.
    * @return a container tag with the submit view
    */
-  public abstract Tag renderFileUploadBlockSubmitForms(
-      Params params, ApplicantQuestionRendererFactory applicantQuestionRendererFactory);
+  public final Tag renderFileUploadBlock(
+      Params params, ApplicantQuestionRendererFactory applicantQuestionRendererFactory) {
+    String onSuccessRedirectUrl =
+        params.baseUrl()
+            + routes.ApplicantProgramBlocksController.updateFile(
+                    params.applicantId(),
+                    params.programId(),
+                    params.block().getId(),
+                    params.inReview())
+                .url();
+
+    String key = FileNameFormatter.formatFileUploadQuestionFilename(params);
+    StorageUploadRequest signedRequest =
+        params.storageClient().getSignedUploadRequest(key, onSuccessRedirectUrl);
+
+    ApplicantQuestionRendererParams rendererParams =
+        ApplicantQuestionRendererParams.builder()
+            .setMessages(params.messages())
+            .setSignedFileUploadRequest(signedRequest)
+            .setErrorDisplayMode(params.errorDisplayMode())
+            .build();
+
+    ContainerTag uploadForm = renderFileUploadFormElement(params, signedRequest);
+    Preconditions.checkState("form".equals(uploadForm.getTagName()), "must be of type form");
+    uploadForm.with(
+        each(
+            params.block().getQuestions(),
+            question ->
+                applicantQuestionRendererFactory.getRenderer(question).render(rendererParams)));
+
+    Tag skipForms = renderDeleteAndContinueFileUploadForms(params);
+    Tag buttons = renderFileUploadBottomNavButtons(params);
+
+    return div(uploadForm, skipForms, buttons).with(each(extraScriptTags(), tag -> footer(tag)));
+  }
+
+  protected ContainerTag renderFileUploadFormElement(Params params, StorageUploadRequest request) {
+    return form()
+        .withId(BLOCK_FORM_ID)
+        .attr(ENCTYPE, "multipart/form-data")
```

Comment:
suggestion (blocking):
```suggestion
        .attr(ENCTYPE, MimeTypes.MULTIPART_FORM_DATA)
```

`MimeTypes` is in `org.eclipse.jetty.http`

Reply:
This would involve bringing in a new build dependency in build.sbt. Just wanted to confirm that it still seems worth it from your perspective.